components/resource_control/src/service.rs

```diff
@@ -278,22 +495,107 @@ pub mod tests {
+        // Set controller config.
+        let cfg = ControllerConfig {
+            request_unit: RequestUnitConfig {
+                read_base_cost: 1. / 8.,
+                read_cost_per_byte: 1. / (64. * 1024.),
+                write_base_cost: 1.,
+                write_cost_per_byte: 1. / 1024.,
+                read_cpu_ms_cost: 1. / 3.,
+            },
+        };
+        store_controller_config(s.clone().pd_client, cfg);
+        let config = block_on(s.load_controller_config());
+        assert_eq!(config.read_base_cost, 1. / 8.);
+
+        server.stop();
+    }
+
+    #[test]
+    fn upload_ru_metrics_test() {
+        let (mut server, client) = new_test_server_and_client(ReadableDuration::millis(100));
+        let resource_manager = ResourceGroupManager::default();
+
+        let s = ResourceManagerService::new(Arc::new(resource_manager), Arc::new(client));
+        let bg = new_background_resource_group_ru("background".into(), 1000, 15, vec!["br".into()]);
+        s.manager.add_resource_group(bg);
+
+        // Set controller config.
+        let cfg = ControllerConfig {
+            request_unit: RequestUnitConfig {
+                read_base_cost: 1. / 8.,
+                read_cost_per_byte: 1. / (64. * 1024.),
+                write_base_cost: 1.,
+                write_cost_per_byte: 1. / 1024.,
+                read_cpu_ms_cost: 1. / 3.,
+            },
+        };
+        store_controller_config(s.clone().pd_client, cfg);
+
+        fail::cfg("set_upload_duration", "return(10)").unwrap();
+        let background_worker = Builder::new("background").thread_count(1).create();
+        let s_clone = s.clone();
+        background_worker.spawn_async_task(async move {
+            s_clone.upload_ru_metrics().await;
+        });
+        // Mock consume.
+        let bg_limiter = s.manager.get_resource_limiter("background", "br").unwrap();
+        bg_limiter.consume(
+            Duration::from_secs(2),
+            IoBytes {
+                read: 1000,
+                write: 1000,
+            },
+        );
+        // Wait for upload ru metrics.
+        std::thread::sleep(Duration::from_millis(100));
+        // Mock update version.
+        let bg = new_resource_group_ru("background".into(), 1000, 15);
+        s.manager.add_resource_group(bg);
+
+        let background_group =
+            new_background_resource_group_ru("background".into(), 500, 8, vec!["lightning".into()]);
+        s.manager.add_resource_group(background_group);
+        let new_bg_limiter = s
+            .manager
+            .get_resource_limiter("background", "lightning")
+            .unwrap();
+        new_bg_limiter.consume(
+            Duration::from_secs(5),
+            IoBytes {
+                read: 2000,
+                write: 2000,
+            },
+        );
+        // Wait for upload ru metrics.
+        std::thread::sleep(Duration::from_millis(100));
```

Comment:
How can this case ensure the metrics are reported to pd, I don't see the test point of this case